File: Pattern/Fast & Slow Pointers/202. Happy Number/Solution.java
class Solution {
    public boolean isHappy(int n) {
        int slow = square(n);
        int fast = square(square(n));
        while(true){
           slow = square(slow);
           fast = square(square(fast));
            if(slow==fast){
                break;
            }
        }
        return slow==1;
    }
    public int square(int n){
        int sum =0;
        int remainder=0;
        while(n>9){
            remainder = n%10;
            sum +=remainder*remainder;
            n =n/10;
        }
        sum+=n*n;
        return sum;
    }
}

File: pattern/Fast & Slow Pointers/202. Happy Number/Solution.py
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = n
        fast = n
        while True:
            slow = self.square(slow)
            fast = self.square(self.square(fast))
            if slow == fast:
                break
        return slow == 1

    def square(self,n:int) -> int:
        sumValue =  0 
        remainder = 0 
        while n>9:
            remainder = n % 10
            sumValue = sumValue + ( remainder * remainder)
            n = n//10
        sumValue = sumValue + (n*n)
        return sumValue